lib/bolt/node/local.rb

```diff
@@ -0,0 +1,117 @@
+require 'json'
+require 'open3'
+require 'bolt/node/result'
+
+module Bolt
+  class Local < Node
+    def connect
+      # this is a local connection, just passing there...
+    end
+
+    def disconnect
+      # this is a local connection, just passing there...
+    end
+
+    def execute(command, options = {})
+      result_output = Bolt::Node::ResultOutput.new
+      status = {}
+
+      @logger.debug { "Executing: #{command}" }
+
+      if options[:stdin]
+        stdout, stderr, rc = Open3.capture3(command, :stdin_data => options[:stdin])
+      else
+        stdout, stderr, rc = Open3.capture3(command)
+      end
+
+      result_output.stdout << stdout.strip unless stdout.nil?
+      @logger.debug { "stdout: #{data}" }
+
+      result_output.stderr << stderr.strip unless stderr.nil?
+      @logger.debug { "stderr: #{data}" }
+
+      status[:exit_code] = rc.to_i
+
+      if status[:exit_code].zero?
+        @logger.debug { "Command returned successfully" }
+        Bolt::Node::Success.new(result_output.stdout.string, result_output)
+      else
+        @logger.info { "Command failed with exit code #{status[:exit_code]}" }
+        Bolt::Node::Failure.new(status[:exit_code], result_output)
+      end
+    end
+
+    def _upload(source, destination)
+      FileUtils.copy_file(source, destination)
+      Bolt::Node::Success.new
+    rescue StandardError => e
+      Bolt::Node::ExceptionFailure.new(e)
+    end
+
+    def make_tempdir
+      # Note(riton): Should be developed to support os that does not have mktemp
+      Bolt::Node::Success.new(`mktemp -d`.chomp)
+    rescue StandardError => e
+      Bolt::Node::ExceptionFailure.new(e)
+    end
+
+    def with_remote_file(file)
+      local_path = ''
+      dir = ''
+      result = nil
+
+      make_tempdir.then do |value|
+        dir = value
+        local_path = "#{dir}/#{File.basename(file)}"
+        Bolt::Node::Success.new
+      end.then do
+        _upload(file, local_path)
+      end.then do
+        execute("chmod u+x '#{local_path}'")
```

Comment:
This will lead to issues if the execution host is _windows_.

Reply:
The winrm transport may provide some hints, and Ruby has methods to detect Windows or non-Windows OS. `Gem.win_platform?` seems reasonable